# Análise do bancos de dados construído pelo Crawler Mercado Livre

Esta é uma análise do banco de dados construído pelo Crawler Mercado Livre. Estes dados foram extraídos durante os dias 12/04/2022 a 21/04/2022 e armazenados num banco postgresql no Heroku. Para mais informações, leia o README.md deste repositório.

## Importando módulos e lendo o banco de dados

##### Módulos

In [616]:
import os
from sqlalchemy import create_engine
import pandas as pd

##### Carregando variáveis de ambiente e lendo o banco de dados como um dataframe

In [617]:
DB_URL = os.environ["DB_URL"].replace("postgres://", "postgresql://", 1)
engine = create_engine(DB_URL)

In [618]:
df = pd.read_sql("SELECT * FROM mercado_livre_offers", con=engine)
pd.options.mode.chained_assignment = None 

## Data wrangling

### Análise inicial

##### Amostra inicial

In [619]:
df.head()

name  \
0  Samsung Galaxy A32 Dual SIM 128 GB branco 4 GB...   
1    Philco Hit P10 Dual SIM 128 GB dourado 4 GB RAM   
2    Cloro Granulado 3 Em 1 Multi Ação 10kg  - Genco   
3  Notebook Samsung Chromebook XE310XBA cinza 11....   
4          Multilaser F Dual SIM 32 GB café 1 GB RAM   

                                          categories  \
0  [Celulares e Telefones, Celulares e Smartphone...   
1  [Celulares e Telefones, Celulares e Smartphone...   
2  [..., ..., ..., Limpeza e Manutenção, Acessóri...   
3  [Informática, Portáteis e Acessórios, Notebook...   
4  [Celulares e Telefones, Celulares e Smartphone...   

                                         best_seller   offer_of_the_day  \
0  [MAIS VENDIDO, 17º em Celulares e Smartphones ...               True   
1                                                 []               True   
2        [MAIS VENDIDO, 1º em Kits Teste Ph e Cloro]              False   
3                                                 []               True   
4                                                 []               True   

  discount old_price new_price       date  
0       15     1.799      1512 2022-04-12  
1       29     1.099       779 2022-04-12  
2        5       301    286.14 2022-04-12  
3       18     2.299      1879 2022-04-12  
4        9       459    414.90 2022-04-12

In [620]:
df.tail()

name  \
9992  Pneu Aro 14 235/60r14 96h Hankook Dynamic Ra03...   
9993  Novo Alisador Cabelo Escova De Barba De Alisam...   
9994  Lipotramina Original Revendedor Autorizado - P...   
9995  Kit 3 Canetas Pintura Porcelana Vidro Cerâmica...   
9996  Caneta Fluorescente Painel Lousa Led - Kit 8 U...   

                                             categories  \
9992  [Acessórios para Veículos, Pneus, Pneus de Car...   
9993  [Beleza e Cuidado Pessoal, Artefatos para Cabe...   
9994                   [Saúde, Suplementos Alimentares]   
9995  [..., Materiais Escolares, Escolar, Insumos pa...   
9996  [..., Materiais Escolares, Escolar, Insumos pa...   

                                            best_seller   offer_of_the_day  \
9992  [MAIS VENDIDO, 11º em Pneus de Carros e Caminh...              False   
9993                                                 []              False   
9994                                                 []               True   
9995         [MAIS VENDIDO, 7º em Marcadores Compactor]              False   
9996                                                 []              False   

     discount old_price new_price       date  
9992       10     1.770      1593 2022-04-21  
9993       45       178     98.02 2022-04-21  
9994       51       202        97 2022-04-21  
9995        5        92     87.87 2022-04-21  
9996        5        78     74.95 2022-04-21

##### Tamanho, colunas e tipos de dados

In [621]:
print(df.shape, '\n', '', '\n', df.columns, '\n', '', '\n', df.dtypes)

(9997, 8) 
  
 Index(['name', 'categories', 'best_seller', ' offer_of_the_day', 'discount',
       'old_price', 'new_price', 'date'],
      dtype='object') 
  
 name                         object
categories                   object
best_seller                  object
 offer_of_the_day              bool
discount                     object
old_price                    object
new_price                    object
date                 datetime64[ns]
dtype: object


##### Verificando valores nulos

In [622]:
df.isnull().sum()

name                   0
categories             0
best_seller            0
 offer_of_the_day      0
discount             300
old_price             46
new_price             76
date                   0
dtype: int64

Aparentemente, os valores nulos estão nas colunas 'discount', 'old_price' e 'new_price', no entanto, na amostra do dataset, encontramos listas vazias na coluna 'best_seller'

### Tratando os dados nulos e erros de cálculos

##### Removendo valores

Vamos verificar os valores nulos nas colunas 'new_price' e 'old_price'. Caso os valores de ambas as colunas sejam nulos ou os valores de 'discount' e 'old_price' ou 'new_price', então não é possível calcular qualquer destes campos

In [623]:
df.loc[df['old_price'].isnull() & df['new_price'].isnull()]

name categories  \
873              LG K62+ Dual SIM 128 GB white 4 GB RAM         []   
2296  Bebida Vegetal Original Orgânico A Tal da Cast...         []   
2518   Bebida Vegetal Amêndoa A Tal da Castanha 1 Litro         []   
2596  Talco Leite de Rosas Barla Perfumado em frasco...         []   
2723  Som automotivo Svart S500 com USB, bluetooth e...         []   
2982    Chips de Pipoca Manteiga Roots To Go Pacote 35g         []   
3611  Fórmula infantil em pó Nestlé Nan Supreme 1  e...         []   
3665    Samsung Galaxy Z Flip3 5G 128 GB cream 8 GB RAM         []   
3670  Samsung Galaxy Note20 Ultra 5G Dual SIM 256 GB...         []   
3903  Cafeteira Cadence O’clock CAF205 semi automáti...         []   
4012  Samsung Galaxy A52 Dual SIM 128 GB branco 6 GB...         []   
4126         Moto E7 Dual SIM 32 GB aquamarine 2 GB RAM         []   
4135  Alto-falante JBL Go 2 portátil com bluetooth m...         []   
4160  Som automotivo Roadstar RS-2709 com USB, bluet...         []   
4630  Notebook Asus VivoBook X543UA cinza-oscura 15....         []   
5105       Máquina de solda WWSoldas MIG 130A azul 220V         []   
5196  Lavadora de alta pressão Kärcher Home & Garden...         []   
5623            Composto Lácteo Fibras Molico Lata 260g         []   
5793  Liquidificador portátil Oster Super Chef BLSTM...         []   
5886  Bicicleta  infantil Nathor Mechanic aro 12 fre...         []   
5954  Xiaomi Pocophone Poco M3 Pro 5G Dual SIM 128 G...         []   
5955       Sabão líquido Mon Bijou Pureza frasco 500 ml         []   
6001  Jack Daniel's Old No. 7 Estados Unidos da Amér...         []   
6164  Cadeira de escritório Waw Design Escritório st...         []   
6207             Fraldas Huggies Tripla Proteção M 42 u         []   
6313       Secador de cabelo Taiff New Black preto 220V         []   
6599  Antitranspirante em barra Nivea Men Clinical I...         []   
6645  Geladeira inverter frost free Panasonic NR-BB5...         []   
6650  Lenço Umedecido Recém-Nascido sem Fragrância J...         []   
6729  Mountain bike TK3 Track TB 100XS aro 26 18" 21...         []   
7202    Fone de ouvido over-ear sem fio CAT STN-28 rosa         []   
7479                      Olympikus Index Homem Adultos         []   
7641                       iPhone 8 Plus 64 GB prateado         []   
7705  Fogão cooktop gás Philco Cook chef 4 preto 110...         []   
7999  Samsung Galaxy S22 Ultra Dual SIM 256 GB phant...         []   
8022  Torneira de cozinha monocomando Lorben Gourmet...         []   
8045  Tablet  Samsung Galaxy Tab A7 Lite SM-T220 8.7...         []   
8352  Cortador de cabelo MQ Hair Gold Series Force B...         []   
8416  Aparelho medidor de pressão arterial digital d...         []   
8445  Fogão cooktop gás Philco Cook chef 5 preto 110...         []   
8661  Cadeira de escritório Trevalla TL-CDG-07 gamer...         []   
8716  Luvas descartáveis Supermax Powder free cor az...         []   
9109  Fogão  de piso Dako Supreme Glass  5Q  gás eng...         []   
9219        Fone de ouvido over-ear AKG K52 matte black         []   
9361  Cortador de grama elétrico Trapp MC 60E de 250...         []   
9421  Capacete Fechado Pro Tork Evolution G8 Evo Bri...         []   

     best_seller   offer_of_the_day discount old_price new_price       date  
873           []              False     None      None      None 2022-04-12  
2296          []              False     None      None      None 2022-04-13  
2518          []              False     None      None      None 2022-04-13  
2596          []              False     None      None      None 2022-04-13  
2723          []              False     None      None      None 2022-04-14  
2982          []              False     None      None      None 2022-04-13  
3611          []              False     None      None      None 2022-04-14  
3665          []              False     None      None      None 2022-04-13  
3670          []              False     None      None    

Vemos que neste caso todos os campos estão sem valores, então vamos remover todos os registros com valores nulos.

In [624]:
df.drop(df.loc[df['old_price'].isnull() & df['new_price'].isnull()].index, inplace=True)

Ainda verificando outros valores nulos

In [625]:
df.loc[df['new_price'].isnull()]

name categories  \
1709           Controle joystick Leboss LB-2121/S preto         []   
2040  Martelete DeWalt D25133K amarelo e preto frequ...         []   
2637  Torneira de cozinha monocomando Pingoo.Casa En...         []   
2745  No break JBR 800 800VA entrada e saída de 120V...         []   
3067  Lixadeira   orbital Mondial FLO-2  preta, verm...         []   
3308  Mouse sem fio Logitech  Pebble M350 cinza-azulado         []   
3479  Teclado sem fio Logitech K230 QWERTY português...         []   
3598   Chivas Regal Blended Extra 13 Reino Unido 750 mL         []   
3767  Caminhonete de controle remoto Importway BW026...         []   
3971  Alto-falante Grasep D-P13 com bluetooth inglat...         []   
4346  Cafeteira Nescafé Arno Dolce Gusto Genio S Bas...         []   
4389  Cafeteira portátil Nescafé Arno Dolce Gusto Mi...         []   
5310  Grand Theft Auto V Standard Edition Rockstar G...         []   
5596  Positivo Twist Mini S431 Dual SIM 8 GB dourado...         []   
5829       Papéis toalha Snob Folha Dupla pacote de 2 u         []   
6115  Macarrão de Sêmola Grano Duro Fusilli 98 Baril...         []   
6214  Fogão  de piso Consul  CFS5QAB  gás engarrafad...         []   
6380  Piscina estrutural retangular Mor 001000 com c...         []   
6944  Ventilador de mesa e parede Mondial V-40 preto...         []   
7386  Furadeira sem fio de 10mm DeWalt DCD700LC1 ama...         []   
7447  Pneu traseira para moto Pirelli Angel ST para ...         []   
7816  Suplemento em  softgels NOW Foods  Vitamin D3 ...         []   
7918  Vinho tinto meio seco Cabernet Sauvignon Trapi...         []   
8007  Tablet  Philco PTB8RSG 8" 32GB cinza e 2GB de ...         []   
8121  Ar condicionado Consul de janela frio 7500 BTU...         []   
8389  Luvas descartáveis Talge Vinil sem pó cor tran...         []   
8530     Aspirador de pó Ultra AP-39 branco e azul 127V         []   
8680  Xiaomi Redmi Note 10 Pro (Global) Dual SIM 128...         []   
8709  Ar condicionado LG split inverter frio 12000 B...         []   
9233       Condicionador Ritual de Reparação Dove 400ml         []   

     best_seller   offer_of_the_day discount old_price new_price       date  
1709          []              False     None        35      None 2022-04-12  
2040          []              False     None       760      None 2022-04-15  
2637          []              False     None       349      None 2022-04-13  
2745          []              False     None       469      None 2022-04-13  
3067          []              False     None       182      None 2022-04-13  
3308          []              False     None        99      None 2022-04-13  
3479          []              False     None       116      None 2022-04-13  
3598          []              False     None       131      None 2022-04-14  
3767          []              False     None       102      None 2022-04-14  
3971          []              False     None        94      None 2022-04-13  
4346          []              False     None       422      None 2022-04-14  
4389          []              False     None       395      None 2022-04-14  
5310          []              False     None       170      None 2022-04-14  
5596          []              False     None       407      None 2022-04-15  
5829          []              False     None         5      None 2022-04-20  
6115          []              False     None         9      None 2022-04-15  
6214          []              False     None     1.290      None 2022-04-15  
6380          []              False     None       113      None 2022-04-15  
6944          []              False     None       143      None 2022-04-19  
7386          []              False     None       493      None 2022-04-16  
7447          []              False     None       717      None 2022-04-19  
7816          []              False     None       107      None 2022-04-18  
7918          []              False     None        49      None 2022-04-19  
8007    

Neste caso, apenas os campos de 'old_price' possuem valores, tornando-se também inviável para qualquer cálculo

In [626]:
df.drop(df.loc[df['new_price'].isnull()].index, inplace=True)

Vamos verificar novamente os valores nulos do nosso dataset

In [627]:
df.isnull().sum()

name                   0
categories             0
best_seller            0
 offer_of_the_day      0
discount             224
old_price              0
new_price              0
date                   0
dtype: int64

##### Transformando o tipo dos dados

Agora vamos transformar os tipos de dados dos campos 'new_price' e 'old_price' para float, para podermos calcular os valores nulos ainda restante na coluna 'discount'.

In [628]:
# Precisei usar o replace pois alguns valores como 1.100 foi convertido para 1.1 em float ao invés de 1100
df['old_price'] = df['old_price'].apply(lambda x: int(x.replace('.', '')))
df['new_price'] = df['new_price'].astype('float')

Vamos transformar também a coluna 'discount' em valores inteiros

In [629]:
df['discount'] = df['discount'].apply(lambda x: int(x) if x != None else x)

##### Corrigindo valores não nulos de 'new_price'

Vemos abaixo que possuimos erros de conta em 'new_price', sendo assim, vamos corrigi-los. 

1) Vamos primeiramente identificar estas linhas

In [630]:
df.loc[(df['old_price'] < df['new_price']) & df['discount'].notnull()]

name  \
470   Sabonete Infantil Em Barra Cremer Loção Hidrat...   
1157  Lençol Descartavel Papel Hospitalar Maca 70x50...   
1553  Leite Longa Vida Semidesnatado com tampa Itala...   
1562               Refrigerante Coca-Cola Garrafa 250ml   
1699  Cabo Otg Usb Tablet Celular Motorola Samsung P...   
...                                                 ...   
8914        50 Escovinha Alongamento Cílios Sobrancelha   
8917      Biscoito Recheio Leite Passatempo Pacote 130g   
9006                       Marshmallow Fini Torção 250g   
9122  Ralo Tampa Grelha Inox Redondo 10 Cm Abre E Fe...   
9153             Sabonete Barra Original Dove Caixa 90g   

                                             categories  \
470   [Beleza e Cuidado Pessoal, Higiene Pessoal, Sa...   
1157  [Beleza e Cuidado Pessoal, Higiene Pessoal, Le...   
1553  [Alimentos e Bebidas, Frescos, Laticínios, Lei...   
1562      [Alimentos e Bebidas, Bebidas, Refrigerantes]   
1699  [Celulares e Telefones, Acessórios para Celula...   
...                                                 ...   
8914  [Beleza e Cuidado Pessoal, Maquiagem, Olhos, C...   
8917        [Alimentos e Bebidas, Mercearia, Biscoitos]   
9006  [Alimentos e Bebidas, Mercearia, Doces e Choco...   
9122  [Casa, Móveis e Decoração, Banheiros, Acessóri...   
9153  [Beleza e Cuidado Pessoal, Higiene Pessoal, Sa...   

                              best_seller   offer_of_the_day  discount  \
470                                    []              False      24.0   
1157                                   []              False       5.0   
1553                                   []              False       6.0   
1562  [MAIS VENDIDO, 6º em Refrigerantes]              False      14.0   
1699    [MAIS VENDIDO, 1º em USB It-Blue]               True       5.0   
...                                   ...                ...       ...   
8914                                   []              False       5.0   
8917     [MAIS VENDIDO, 15º em Biscoitos]              False      10.0   
9006                                   []              False       7.0   
9122        [MAIS VENDIDO, 6º em Grelhas]              False       7.0   
9153          [MAIS VENDIDO, 1º em Barra]              False      13.0   

      old_price  new_price       date  
470           3       3.03 2022-04-12  
1157         10      10.35 2022-04-12  
1553          4       4.12 2022-04-13  
1562          3       3.34 2022-04-13  
1699         13      13.19 2022-04-13  
...         ...        ...        ...  
8914         18      18.04 2022-04-20  
8917          2       2.24 2022-04-20  
9006          7       7.43 2022-04-20  
9122         11      11.15 2022-04-20  
9153          5       5.09 2022-04-20  

[66 rows x 8 columns]

2) Vamos adicionar este dataframe a uma variável e remove-lo do dataset original

In [631]:
df_fix_new_price = df.loc[(df['old_price'] < df['new_price']) & df['discount'].notnull()]
df.drop(df.loc[(df['old_price'] < df['new_price']) & df['discount'].notnull()].index, inplace=True)

3) Agora vamos corrigir os valores de 'new_price'

In [632]:
df_fix_new_price['new_price'] = df_fix_new_price['old_price']*(100-df_fix_new_price['discount'])/100

4. Agora vamos adicionar este dataframe ao dataset original

In [633]:
df = pd.concat([df, df_fix_new_price])

##### Corrigindo valores nulos de discount

Vamos realiazar as mesmas operações, separar, remover, corrigir e adicionar os valores novamente no dataframe

In [634]:
df_discount_null = df.loc[df['discount'].isnull()]
df.drop(df.loc[df['discount'].isnull()].index, inplace=True)

In [635]:
df_discount_null['discount'] = round((1-(df_discount_null['new_price']/df_discount_null['old_price']))*100, 2)

In [636]:
df_discount_null.head()

name  \
70      Micro-ondas Electrolux MTD30   branco 20L 127V   
257  Placa Interface Geladeira Electrolux Df50 6480...   
297              Cabo Flexível 10,0mm Preto 100m - Sil   
318  Maleta Organizadora 34 Esmaltes Manicure+porta...   
343  Tablet  Samsung Galaxy Tab S6 Lite SM-P615 10....   

                                            categories  \
70   [Eletrodomésticos, Fornos e Fogões, Microondas...   
257  [Eletrodomésticos, Refrigeração, Peças e Acess...   
297  [Construção, Energia, Cabos e Acessórios, Filt...   
318  [Beleza e Cuidado Pessoal, Manicure e Pedicure...   
343  [Informática, Tablets e Acessórios, Tablets, S...   

                                      best_seller   offer_of_the_day  \
70              [MAIS VENDIDO, 19º em Microondas]               True   
257  [MAIS VENDIDO, 2º em Placas para Geladeiras]              False   
297                                            []              False   
318         [MAIS VENDIDO, 1º em Esmaltes Dompel]               True   
343                 [MAIS VENDIDO, 1º em Tablets]              False   

     discount  old_price  new_price       date  
70       3.35        537     519.00 2022-04-12  
257     -0.48        199     199.95 2022-04-12  
297      4.91        939     892.91 2022-04-12  
318      3.44         61      58.90 2022-04-12  
343      4.80       2499    2379.00 2022-04-12

Alguns valores de 'discount' ficaram negativos devido aos valores de new_price serem maior do que old_price. Vamos remover estes registros visto que não faz sentido um desconto negativo, aumentando o preço final.

In [637]:
df_discount_null.loc[df_discount_null['discount'] < 0].head()

name  \
257   Placa Interface Geladeira Electrolux Df50 6480...   
380   Lixa De Polir Unhas Bloco Branco 4 Faces Brilh...   
435   Suporte Lixa Pés Anatômica Lixa Descartável Re...   
999   Acabamento Quadrado Para Registro Padrão Deca ...   
1089  Driver  D220ti Trio Titanio - 90 Watts Rms  Jb...   

                                             categories  \
257   [Eletrodomésticos, Refrigeração, Peças e Acess...   
380   [Beleza e Cuidado Pessoal, Manicure e Pedicure...   
435   [Beleza e Cuidado Pessoal, Manicure e Pedicure...   
999   [Construção, Mobiliário para Banheiros, Tornei...   
1089  [Acessórios para Veículos, Som Automotivo, Dri...   

                                         best_seller   offer_of_the_day  \
257     [MAIS VENDIDO, 2º em Placas para Geladeiras]              False   
380      [MAIS VENDIDO, 6º em Lixas de Unhas Dompel]               True   
435     [MAIS VENDIDO, 20º em Lixas de Unhas Dompel]               True   
999   [MAIS VENDIDO, 17º em Torneiras Convencionais]              False   
1089                                              []              False   

      discount  old_price  new_price       date  
257      -0.48        199     199.95 2022-04-12  
380     -10.57          7       7.74 2022-04-12  
435     -11.86          7       7.83 2022-04-12  
999      -1.96         46      46.90 2022-04-12  
1089     -0.18        248     248.45 2022-04-15

In [638]:
df_discount_null.drop(df_discount_null.loc[df_discount_null['discount'] < 0].index, inplace=True)

Agora vamos adicionar este dataframe ao dataset original novamente

In [639]:
df = pd.concat([df, df_discount_null])

Verificando se ainda há valores nulos

In [640]:
df.isnull().sum()

name                 0
categories           0
best_seller          0
 offer_of_the_day    0
discount             0
old_price            0
new_price            0
date                 0
dtype: int64

##### Calculando valor real de 'discount'

Fazendo uma conta simples, podemos perceber que o valor de 'discount' mostrado no dataset não é valor real de desconto do 'old_price' para o 'new_price'.

In [641]:
df.head(1)

name  \
0  Samsung Galaxy A32 Dual SIM 128 GB branco 4 GB...   

                                          categories  \
0  [Celulares e Telefones, Celulares e Smartphone...   

                                         best_seller   offer_of_the_day  \
0  [MAIS VENDIDO, 17º em Celulares e Smartphones ...               True   

   discount  old_price  new_price       date  
0      15.0       1799     1512.0 2022-04-12

Vamos calcular o desconto deste primeiro produto

In [642]:
(1-1512/1799)*100

15.953307392996107

Vemos que temos uma diferença de quase 1% entre o valor no dataset e o valor real de desconto. Desta maneira, vamos recalcular todos os descontos dos produtos.

In [643]:
df['discount'] = round((1-(df['new_price']/df['old_price']))*100, 2)

In [644]:
df.head()

name  \
0  Samsung Galaxy A32 Dual SIM 128 GB branco 4 GB...   
1    Philco Hit P10 Dual SIM 128 GB dourado 4 GB RAM   
2    Cloro Granulado 3 Em 1 Multi Ação 10kg  - Genco   
3  Notebook Samsung Chromebook XE310XBA cinza 11....   
4          Multilaser F Dual SIM 32 GB café 1 GB RAM   

                                          categories  \
0  [Celulares e Telefones, Celulares e Smartphone...   
1  [Celulares e Telefones, Celulares e Smartphone...   
2  [..., ..., ..., Limpeza e Manutenção, Acessóri...   
3  [Informática, Portáteis e Acessórios, Notebook...   
4  [Celulares e Telefones, Celulares e Smartphone...   

                                         best_seller   offer_of_the_day  \
0  [MAIS VENDIDO, 17º em Celulares e Smartphones ...               True   
1                                                 []               True   
2        [MAIS VENDIDO, 1º em Kits Teste Ph e Cloro]              False   
3                                                 []               True   
4                                                 []               True   

   discount  old_price  new_price       date  
0     15.95       1799    1512.00 2022-04-12  
1     29.12       1099     779.00 2022-04-12  
2      4.94        301     286.14 2022-04-12  
3     18.27       2299    1879.00 2022-04-12  
4      9.61        459     414.90 2022-04-12

Agora vamos verificar se não temos nenhum valor negativo de desconto

In [645]:
df.loc[df['discount'] < 0]

Empty DataFrame
Columns: [name, categories, best_seller,  offer_of_the_day, discount, old_price, new_price, date]
Index: []

### Criando novas colunas

#### Definido os valores para as colunas specific_category e category

Como podemos ver abaixo, cada produto se encaixa em diversas categorias, as quais vão se especificando mais e mais dentro do site do mercado livre. Deste modo, vamos criar duas novas colunas:

1. Coluna 'category', a qual representa uma categoria mais geral do produto
2. Coluna 'specific_category', a qual representa uma categoria específica do produto com base na coluna 'best_seller'

In [646]:
df['categories'].head()

0    [Celulares e Telefones, Celulares e Smartphone...
1    [Celulares e Telefones, Celulares e Smartphone...
2    [..., ..., ..., Limpeza e Manutenção, Acessóri...
3    [Informática, Portáteis e Acessórios, Notebook...
4    [Celulares e Telefones, Celulares e Smartphone...
Name: categories, dtype: object

In [647]:
df['best_seller'].head()

0    [MAIS VENDIDO, 17º em Celulares e Smartphones ...
1                                                   []
2          [MAIS VENDIDO, 1º em Kits Teste Ph e Cloro]
3                                                   []
4                                                   []
Name: best_seller, dtype: object

#### coluna specific_category

##### Definindo coluna specific_category

Podemos notar que em 'best_seller' temos uma lista com dois valores. O primeiro se trata de anunciar que aquele produto estava entre os mais vendidos daquele dia, e o segundo é a categoria que se encontrava. Deste modo, vamos definir a categoria específica com base no segundo valor desta lista.

Precisamos de uma função que faça igual o procedimento abaixo: 

In [648]:
categoria = ['MAIS VENDIDO','14º em Máquinas de Cortar Cabelo Wmark']
' '.join(categoria[1].split()[2:])

'Máquinas de Cortar Cabelo Wmark'

In [649]:
df['specific_category'] = df['best_seller'].apply(lambda x: ' '.join(x[1].split()[2:]) if x != [] else '')

##### Resultado e análise da coluna specific_category

Vamos ver o dataframe com a nova coluna

In [650]:
df.head(10)

name  \
0  Samsung Galaxy A32 Dual SIM 128 GB branco 4 GB...   
1    Philco Hit P10 Dual SIM 128 GB dourado 4 GB RAM   
2    Cloro Granulado 3 Em 1 Multi Ação 10kg  - Genco   
3  Notebook Samsung Chromebook XE310XBA cinza 11....   
4          Multilaser F Dual SIM 32 GB café 1 GB RAM   
5  Módulo Taramps Ts400 Digital 400w Rms 4 Canais...   
6          Moto G31 Dual SIM 128 GB grafite 4 GB RAM   
7   Samsung Galaxy A71 Dual SIM 128 GB azul 6 GB RAM   
8  Fritadeira Air Fry Inox Bfr11pi 3,5 Litros Bri...   
9              Moto E20 Dual SIM 32 GB azul 2 GB RAM   

                                          categories  \
0  [Celulares e Telefones, Celulares e Smartphone...   
1  [Celulares e Telefones, Celulares e Smartphone...   
2  [..., ..., ..., Limpeza e Manutenção, Acessóri...   
3  [Informática, Portáteis e Acessórios, Notebook...   
4  [Celulares e Telefones, Celulares e Smartphone...   
5  [Acessórios para Veículos, Som Automotivo, Mód...   
6  [Celulares e Telefones, Celulares e Smartphone...   
7  [Celulares e Telefones, Celulares e Smartphone...   
8  [..., Pequenos Eletrodomésticos, Para a Cozinh...   
9  [Celulares e Telefones, Celulares e Smartphone...   

                                         best_seller   offer_of_the_day  \
0  [MAIS VENDIDO, 17º em Celulares e Smartphones ...               True   
1                                                 []               True   
2        [MAIS VENDIDO, 1º em Kits Teste Ph e Cloro]              False   
3                                                 []               True   
4                                                 []               True   
5       [MAIS VENDIDO, 1º em Módulos Amplificadores]              False   
6                                                 []               True   
7                                                 []              False   
8                        [MAIS VENDIDO, 1º em De Ar]              False   
9      [MAIS VENDIDO, 5º em Celulares e Smartphones]               True   

   discount  old_price  new_price       date                specific_category  
0     15.95       1799    1512.00 2022-04-12  Celulares e Smartphones Samsung  
1     29.12       1099     779.00 2022-04-12                                   
2      4.94        301     286.14 2022-04-12            Kits Teste Ph e Cloro  
3     18.27       2299    1879.00 2022-04-12                                   
4      9.61        459     414.90 2022-04-12                                   
5     10.54        257     229.90 2022-04-12           Módulos Amplificadores  
6     15.60       1699    1434.00 2022-04-12                                   
7     47.16       3499    1849.00 2022-04-12                                   
8     47.76        689     359.90 2022-04-12                            De Ar  
9     28.21       1099     789.00 2022-04-12          Celulares e Smartphones

Vamos ver quais e quantas categorias específicas nós temos

In [651]:
df['specific_category'].value_counts()

                                5932
Celulares e Smartphones           18
Ar Condicionado                   16
Televisores                       13
Elétricos                         11
                                ... 
Espelhos para Banheiro Astra       1
Panetones                          1
Aplicadores de Tags                1
Balanças G-Tech                    1
Motores de Combustão               1
Name: specific_category, Length: 1997, dtype: int64

Tirando os valores '', temos 1996 categorias específicas. Contudo, ainda temos 5932 valores vazios. Vamos tentar preenche-los.

##### Preenchendo valores vazios de specific_category

Vamos agora preencher os valores vazios da coluna 'specific_category' procurando estes valores dentro da lista que está na coluna 'categories'

In [652]:
# Criando uma lista com os valores a serem procurados

specific = df['specific_category'].value_counts().index
specific = specific.to_list()
specific[1:]

['Celulares e Smartphones',
 'Ar Condicionado',
 'Televisores',
 'Elétricos',
 'Celulares e Smartphones Motorola',
 'Monitores',
 'Descartáveis',
 'Vinhos',
 'Fontes Chaveadas',
 'Desodorantes',
 'Divisores de Frequência',
 'Protetores Solares',
 'Brocas',
 'Computadores',
 'Kit de Chaves',
 'Travesseiros',
 'Sem Tela',
 'Torneiras Convencionais',
 'Cervejas',
 'Secadores de Cabelo',
 'Suplementos',
 'Baterias',
 'Suplementos Alimentares Integralmédica',
 'Lubrificantes',
 'Tapete Eva',
 'Lixadeiras',
 'Andadores',
 'Extensores e Elásticos',
 'Joysticks',
 'Amaciante',
 'Suportes',
 'Balanças',
 'Notebooks',
 'Shampoos e Condicionadores',
 'Barras de Led',
 'Kit de Chaves Sata',
 'Cabides',
 'Jogos e Kits de Toalhas',
 'Grupos e Kits',
 'Televisores Philips',
 'Perfumes',
 'Antenas de TV',
 'Grelhas',
 'Celulares e Smartphones Apple',
 'Scanners',
 'Para TV',
 'Receptores para TV',
 'Alarmes e Sensores',
 'Grampeadores e Pinadores',
 'Geladeiras',
 'Placas Acústicas',
 'Digitais',
 'Fo

Função para procurar valores na lista e preencher os valores nulos de 'specific_category'

In [653]:
def fill_specific_category_empty_values(specific_category, categories, specific):
    if specific_category == '':
        for cat in specific:
            if cat in categories:
                return cat
    elif specific_category != '':
        return specific_category
    else:
        return ''

In [654]:
df['specific_category'] = df.apply(lambda row: fill_specific_category_empty_values(
    row['specific_category'], row['categories'], specific),
    axis=1)

In [655]:
df['specific_category'].value_counts()

Ferramentas                         221
Celulares e Smartphones             161
Livros Fisicos                      153
Tratamentos com o Cabelo            107
Computadores                         73
                                   ... 
Trampolins Vital Esporte              1
Máquinas de Cortar Cabelo Shinon      1
Espumantes e Champagne Freixenet      1
Cafeteiras Bialetti                   1
Motores de Combustão                  1
Name: specific_category, Length: 1996, dtype: int64

In [656]:
df.isnull().sum()

name                    0
categories              0
best_seller             0
 offer_of_the_day       0
discount                3
old_price               0
new_price               0
date                    0
specific_category    1448
dtype: int64

Podemos ver que diversas linhas foram preenchidas, porém ainda temos 1448 em branco.

##### Excluindo alguns valores nulos restantes de discount

Apareceram 3 linhas com valores nulos de desconto e preços zerados. Vamos exclui-los.

In [657]:
df.loc[df['discount'].isnull()]

name  \
2884  Bala Halls mentol extra forte sem glúten 27.5 g    
5440                Bala Halls mentol sem glúten 28 g    
6039                 Bala Halls menta sem glúten 28 g    

                                             categories  \
2884  [Alimentos e Bebidas, Mercearia, Doces e Choco...   
5440  [Alimentos e Bebidas, Mercearia, Doces e Choco...   
6039  [Alimentos e Bebidas, Mercearia, Doces e Choco...   

                            best_seller   offer_of_the_day  discount  \
2884   [MAIS VENDIDO, 4º em Guloseimas]              False       NaN   
5440  [MAIS VENDIDO, 10º em Guloseimas]              False       NaN   
6039   [MAIS VENDIDO, 9º em Guloseimas]              False       NaN   

      old_price  new_price       date specific_category  
2884          0        0.0 2022-04-13        Guloseimas  
5440          0        0.0 2022-04-15        Guloseimas  
6039          0        0.0 2022-04-16        Guloseimas

In [658]:
df.drop(df.loc[df['discount'].isnull()].index, inplace=True)

#### coluna category

##### Criando coluna 'category' com base na coluna 'categories'

In [659]:
df['category'] = df['categories'].apply(lambda x: x[0])

Visualizando a nova coluna

In [660]:
df['category'].head(10)

0       Celulares e Telefones
1       Celulares e Telefones
2                         ...
3                 Informática
4       Celulares e Telefones
5    Acessórios para Veículos
6       Celulares e Telefones
7       Celulares e Telefones
8                         ...
9       Celulares e Telefones
Name: category, dtype: object

##### Consertando valores '...' da coluna category

Algumas linhas ficaram com os valores '...'. Vamos consertar isto

In [661]:
df.loc[df['category'] == '...']

name  \
2       Cloro Granulado 3 Em 1 Multi Ação 10kg  - Genco   
8     Fritadeira Air Fry Inox Bfr11pi 3,5 Litros Bri...   
24    Câmera de segurança Intelbras iM3 com resoluçã...   
34    Camera Ip Lampada Segurança 360 Visão Noturna ...   
44    Alto-falante Philco Go PBS25BT portátil com bl...   
...                                                 ...   
5078  Kit Cartucho Hp 667xl Original Preto+color Par...   
7194  2 Amortecedor Dianteiro Cofap + Coxim Vw Gol 1...   
7703  Capota Strada 2021 2022 2023 Maritima Trek Alu...   
8926  Sanduicheira Grill Antiaderente 800w Inox S-19...   
8974  5l Grande Capacidade Ao Ar Livre Ciclismo Moch...   

                                             categories  \
2     [..., ..., ..., Limpeza e Manutenção, Acessóri...   
8     [..., Pequenos Eletrodomésticos, Para a Cozinh...   
24    [..., ..., Sistemas de Monitoramento, Câmeras ...   
34    [..., Segurança para Casa, Sistemas de Monitor...   
44    [..., ..., Áudio Portátil e Acessórios, Acessó...   
...                                                 ...   
5078  [..., Impressão, Suprimentos de Impressão, Car...   
7194  [..., Peças de Carros e Caminhonetes, Suspensã...   
7703  [..., Aces. de Carros e Caminhonetes, Exterior...   
8926  [..., Pequenos Eletrodomésticos, Para a Cozinh...   
8974  [..., Ciclismo, Acessórios , Transporte e Carg...   

                                      best_seller   offer_of_the_day  \
2     [MAIS VENDIDO, 1º em Kits Teste Ph e Cloro]              False   
8                     [MAIS VENDIDO, 1º em De Ar]              False   
24                                             []              False   
34                                             []               True   
44         [MAIS VENDIDO, 1º em Caixas Bluetooth]              False   
...                                           ...                ...   
5078                                           []              False   
7194                                           []              False   
7703                                           []              False   
8926                                           []              False   
8974                                           []              False   

      discount  old_price  new_price       date      specific_category  \
2         4.94        301     286.14 2022-04-12  Kits Teste Ph e Cloro   
8        47.76        689     359.90 2022-04-12                  De Ar   
24       21.66        319     249.90 2022-04-12   Câmeras de Segurança   
34        9.00        136     123.76 2022-04-12   Câmeras de Segurança   
44       25.13        199     149.00 2022-04-12       Caixas Bluetooth   
...        ...        ...        ...        ...                    ...   
5078      0.00        175     175.00 2022-04-15                   None   
7194      4.79        323     307.52 2022-04-20          Amortecedores   
7703      3.68        535     515.33 2022-04-19                Capotas   
8926      3.70        138     132.90 2022-04-20                   None   
8974      4.94        133     126.43 2022-04-20                   None   

     category  
2         ...  
8         ...  
24        ...  
34        ...  
44        ...  
...       ...  
5078      ...  
7194      ...  
7703      ...  
8926      ...  
8974      ...  

[503 rows x 10 columns]

Função para corrigir os valores '...' de category

In [662]:
def fill_category_values(category, categories):
    if category == '...':
        for cat in categories:
            if cat != '...':
                return cat
    else:
        return category

In [663]:
df['category'] = df.apply(lambda row: fill_category_values(
    row['category'], row['categories']), axis=1)

In [664]:
df.loc[df['category'] == '...']

Empty DataFrame
Columns: [name, categories, best_seller,  offer_of_the_day, discount, old_price, new_price, date, specific_category, category]
Index: []

In [665]:
df['category'].head(10)

0        Celulares e Telefones
1        Celulares e Telefones
2         Limpeza e Manutenção
3                  Informática
4        Celulares e Telefones
5     Acessórios para Veículos
6        Celulares e Telefones
7        Celulares e Telefones
8    Pequenos Eletrodomésticos
9        Celulares e Telefones
Name: category, dtype: object

Pronto, arrumamos todos os valores desta coluna

#### Redefinindo a coluna best_seller

##### Ideia

Como a coluna best_seller nos mostra que este produto estava entre os mais vendidos do dia, sua posição e categoria especifica, vamos alterar os valores desta coluna apenas com a posição de venda deste produto, pois já criamos a coluna specific_category com base nela.

##### execução

In [666]:
def get_position(row):
    if row != []:
        for i in row[1]:
            if i.isnumeric():
                return i
    else:
        return None


df['best_seller'] = df['best_seller'].apply(lambda row: get_position(row))

In [667]:
df.head(10)

name  \
0  Samsung Galaxy A32 Dual SIM 128 GB branco 4 GB...   
1    Philco Hit P10 Dual SIM 128 GB dourado 4 GB RAM   
2    Cloro Granulado 3 Em 1 Multi Ação 10kg  - Genco   
3  Notebook Samsung Chromebook XE310XBA cinza 11....   
4          Multilaser F Dual SIM 32 GB café 1 GB RAM   
5  Módulo Taramps Ts400 Digital 400w Rms 4 Canais...   
6          Moto G31 Dual SIM 128 GB grafite 4 GB RAM   
7   Samsung Galaxy A71 Dual SIM 128 GB azul 6 GB RAM   
8  Fritadeira Air Fry Inox Bfr11pi 3,5 Litros Bri...   
9              Moto E20 Dual SIM 32 GB azul 2 GB RAM   

                                          categories best_seller  \
0  [Celulares e Telefones, Celulares e Smartphone...           1   
1  [Celulares e Telefones, Celulares e Smartphone...        None   
2  [..., ..., ..., Limpeza e Manutenção, Acessóri...           1   
3  [Informática, Portáteis e Acessórios, Notebook...        None   
4  [Celulares e Telefones, Celulares e Smartphone...        None   
5  [Acessórios para Veículos, Som Automotivo, Mód...           1   
6  [Celulares e Telefones, Celulares e Smartphone...        None   
7  [Celulares e Telefones, Celulares e Smartphone...        None   
8  [..., Pequenos Eletrodomésticos, Para a Cozinh...           1   
9  [Celulares e Telefones, Celulares e Smartphone...           5   

    offer_of_the_day  discount  old_price  new_price       date  \
0               True     15.95       1799    1512.00 2022-04-12   
1               True     29.12       1099     779.00 2022-04-12   
2              False      4.94        301     286.14 2022-04-12   
3               True     18.27       2299    1879.00 2022-04-12   
4               True      9.61        459     414.90 2022-04-12   
5              False     10.54        257     229.90 2022-04-12   
6               True     15.60       1699    1434.00 2022-04-12   
7              False     47.16       3499    1849.00 2022-04-12   
8              False     47.76        689     359.90 2022-04-12   
9               True     28.21       1099     789.00 2022-04-12   

                 specific_category                   category  
0  Celulares e Smartphones Samsung      Celulares e Telefones  
1          Celulares e Smartphones      Celulares e Telefones  
2            Kits Teste Ph e Cloro       Limpeza e Manutenção  
3                        Notebooks                Informática  
4          Celulares e Smartphones      Celulares e Telefones  
5           Módulos Amplificadores   Acessórios para Veículos  
6          Celulares e Smartphones      Celulares e Telefones  
7          Celulares e Smartphones      Celulares e Telefones  
8                            De Ar  Pequenos Eletrodomésticos  
9          Celulares e Smartphones      Celulares e Telefones

Os valores None na coluna best_seller representam que o produto não estava entre os mais vendidos do dia e são importantes para a análise.

### Procedimentos finais da limpeza dos dados

Vendo os dados nulos restantes, vamos ignorar a coluna best_seller como descrito anteriormente e preencher os valores faltantes de specific_category com o último elemento da lista na coluna categories.

In [668]:
df.isnull().sum()

name                    0
categories              0
best_seller          5932
 offer_of_the_day       0
discount                0
old_price               0
new_price               0
date                    0
specific_category    1448
category                0
dtype: int64

#### Preenchendo valores faltantes de specific_category

In [671]:
df_specific_cat_none = df.loc[df['specific_category'].isnull()]

In [672]:
df_specific_cat_none['specific_category'] = df_specific_cat_none['categories'].apply(lambda x: x[-1])

##### Removendo do dataset original

In [673]:
df.drop(df.loc[df['specific_category'].isnull()].index, inplace=True)

##### Análise dos dados preenchidos

In [674]:
df_specific_cat_none['specific_category'].value_counts()

Outros                        461
Outros Utensílios              56
King                           39
Solteiro                       36
Outros Produtos de Limpeza     33
                             ... 
Bateria                         1
Fechaduras de Portas            1
Pais e Filhos                   1
Guitarras e Violões             1
Training                        1
Name: specific_category, Length: 470, dtype: int64

Vemos que 470 novas categorias específicas foram adicionadas.

##### Concat dos dos dados no dataframe original

In [681]:
df = pd.concat([df, df_specific_cat_none])

#### Excluindo coluna categories

Após todas as modificações, podemos excluir a coluna categories, pois toda informação está disponível em 'specific_category' e 'category'.

In [687]:
df.drop(columns='categories', axis=0, inplace=True)

#### Definindo os dias da coluna date

Visto que já sabemos que estes dados foram coletados entre os dias 12/04/2022 e 21/04/2022, vamos definir a coluna date apenas com os dias, sem a necessidade do mês, ano e horários.

In [693]:
df['date'] = df['date'].dt.day

#### Amostra final do dataframe

In [696]:
df.head()

name best_seller  \
0  Samsung Galaxy A32 Dual SIM 128 GB branco 4 GB...           1   
1    Philco Hit P10 Dual SIM 128 GB dourado 4 GB RAM        None   
2    Cloro Granulado 3 Em 1 Multi Ação 10kg  - Genco           1   
3  Notebook Samsung Chromebook XE310XBA cinza 11....        None   
4          Multilaser F Dual SIM 32 GB café 1 GB RAM        None   

    offer_of_the_day  discount  old_price  new_price  date  \
0               True     15.95       1799    1512.00    12   
1               True     29.12       1099     779.00    12   
2              False      4.94        301     286.14    12   
3               True     18.27       2299    1879.00    12   
4               True      9.61        459     414.90    12   

                 specific_category               category  
0  Celulares e Smartphones Samsung  Celulares e Telefones  
1          Celulares e Smartphones  Celulares e Telefones  
2            Kits Teste Ph e Cloro   Limpeza e Manutenção  
3                        Notebooks            Informática  
4          Celulares e Smartphones  Celulares e Telefones

In [697]:
df.tail()

name best_seller  \
8926  Sanduicheira Grill Antiaderente 800w Inox S-19...        None   
8974  5l Grande Capacidade Ao Ar Livre Ciclismo Moch...        None   
8994                                   Cajon + Bag Case        None   
9399  Mochila De Hidratação Aonijie Running Ao Ar Livre        None   
9418                     Olympikus Linear Homem Adultos        None   

       offer_of_the_day  discount  old_price  new_price  date  \
8926              False      3.70        138     132.90    20   
8974              False      4.94        133     126.43    20   
8994              False      0.00        399     399.00    19   
9399               True      1.86        162     158.99    21   
9418              False      0.00        179     179.00    21   

     specific_category                   category  
8926           Mondial  Pequenos Eletrodomésticos  
8974            Lixada                   Ciclismo  
8994           Cajones      Instrumentos Musicais  
9399            Outros         Esportes e Fitness  
9418          Training  Calçados, Roupas e Bolsas

In [698]:
df.isnull().sum() # valos nulos necessários na coluna best_seller

name                    0
best_seller          5932
 offer_of_the_day       0
discount                0
old_price               0
new_price               0
date                    0
specific_category       0
category                0
dtype: int64

In [699]:
print(df.shape, '\n', '', '\n', df.columns, '\n', '', '\n', df.dtypes)

(9844, 9) 
  
 Index(['name', 'best_seller', ' offer_of_the_day', 'discount', 'old_price',
       'new_price', 'date', 'specific_category', 'category'],
      dtype='object') 
  
 name                  object
best_seller           object
 offer_of_the_day       bool
discount             float64
old_price              int64
new_price            float64
date                   int64
specific_category     object
category              object
dtype: object


Após todos estes procedimentos de limpeza, podemos partir para uma análise dos dados.

## Data Analysis